### PR TITLE
feat(#6543): adds api-side support for users with multiple facilities

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Covers:
- support for downloading and uploading doc permissions when a user has multiple facilities
- new users v3 api to add a user with multiple facilities
- new users v3 api to update a user to add multiple facilities
- updated users shared library to support multiple facilities 
- updated recent users search api to support multiple facilities

Implementation details:
- when creating a user with multiple facilities, the contact has to be within one of the facilities
- /v3/users accepts the same payload and /v1/users
- /v3/users does not create facilities, only accepts an array of uuids of existing facilities
- /v3/users does not create a contact

Does not cover:
- validation to ensure all facilities are of the same type. We can add this, but right now it's not implemented

#6543 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6543-multi-facility/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6543-multi-facility/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:6543-multi-facility/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Review by latin-panda — DISMISSED** [review]
@dianabarsan this is great! Thanks a lot
I left few comments as I was reading to understand the changes

I tested the app and seems to work fine for users with 1 facility and with more than one:

<details> <summary> contact list</summary>

Screenshot 2024-05-17 at 4 23 47 PM

</details>

<details> <summary> contact details</summary>

Screenshot 2024-05-17 at 4 25 22 PM

</details>

<details> <summary> breadcrumbs </summary>

open in a tab
Screenshot 2024-05-17 at 4 29 47 PM


</details> 

I was able to create and update users:
<details> <summary> Create </summary>

Screenshot 2024-05-17 at 4 21 51 PM


</details> 

When I change or add a facility to the user, my session expires, then I login, I still the see the old stuff, but after sync, I see the facilities updated.

File: api/src/controllers/users.js
Comment (dianabarsan) [reply api/src/controllers/users.js:287]: I believe this is beyond the scope of this PR or this issue :)

File: shared-libs/user-management/src/users.js
Comment (dianabarsan) [reply shared-libs/user-management/src/users.js:690]: I'm not sure. I did do a bunch of sonar-requested refactors, but only for functions that I actually touched. I think it's annoying too.

File: tests/e2e/default/contacts/edit-person-home-place.wdio-spec.js
Comment (dianabarsan) [reply tests/e2e/default/contacts/edit-person-home-place.wdio-spec.js:9]: This is on purpose. I think the frontend work that you are doing will need to handle this skip. There's no point in making this e2e test work, just so you edit it in the next pr,.

File: tests/e2e/default/contacts/person-under-area.wdio-spec.js
Comment (dianabarsan) [reply tests/e2e/default/contacts/person-under-area.wdio-spec.js:48]: This is on purpose. I think the frontend work that you are doing will need to handle this skip. There's no point in making this e2e test work, just so you edit it in the next pr,.

File: api/src/controllers/users.js
Comment (latin-panda) [reply api/src/controllers/users.js:287]: should we create a ticket to follow up on bulk upload later for v3?

File: tests/e2e/default/contacts/person-under-area.wdio-spec.js
Comment (latin-panda) [reply tests/e2e/default/contacts/person-under-area.wdio-spec.js:48]: Ah got it, we'll make a PR to fix those e2e

File: api/src/controllers/users.js
Comment (dianabarsan) [reply api/src/controllers/users.js:287]: I think that's premature to decide. I don't know any instances of anyone using bulk user upload, and the user management tool is definitely not using it. The ticket would be about _discussing_ adding bulk upload.

File: api/src/controllers/users.js
Comment (dianabarsan) [reply api/src/controllers/users.js:290]: So this endpoint will throw an error if the role lacks the multi-facility permission but multiple facilities are sent?

File: shared-libs/user-management/src/users.js
Comment (dianabarsan) [reply shared-libs/user-management/src/users.js:413]: I think we actually want to update the `data.roles` property.

File: api/src/controllers/users.js
Comment (latin-panda) [reply api/src/controllers/users.js:290]: Yes, if roles lack multi-facility permission but receive a user with 1 facility, then it's successful. If it receives many facilities then throws an error

File: shared-libs/user-management/src/users.js
Comment (latin-panda) [reply shared-libs/user-management/src/users.js:413]: Oh why?

File: tests/e2e/default/targets/target-aggregates.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/default/targets/target-aggregates.wdio-spec.js:73]: This e2e is already fixed in this other [PR](https://github.com/medic/cht-core/pull/9099#partial-pull-merging)

File: tests/integration/api/controllers/replication.spec.js
Comment (latin-panda) [inline tests/integration/api/controllers/replication.spec.js:1070]: Same. All these tests are already working, and no extra changes are required. [I enabled them in a branch](https://github.com/medic/cht-core/actions/runs/9175810540).

**Review by Benmuiruri — COMMENTED** [review]
We need to update the create-user [URL](https://github.com/medic/cht-core/blob/fc98a76f9128b4737a366f210a1330229529474a/admin/src/js/services/create-user.js#L4) to point to `v3`

**Comment by latin-panda** [comment]
@Benmuiruri, I think you can use v3 for single user, and keep v2 for multiple users.

File: tests/e2e/default/contacts/edit-person-home-place.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/default/contacts/edit-person-home-place.wdio-spec.js:9]: @Benmuiruri fixed it in [this PR](https://github.com/medic/cht-core/pull/9128).

File: tests/e2e/default/contacts/person-under-area.wdio-spec.js
Comment (latin-panda) [inline tests/e2e/default/contacts/person-under-area.wdio-spec.js:48]: @Benmuiruri fixed it in [this PR](https://github.com/medic/cht-core/pull/9128).

File: shared-libs/user-management/src/users.js
Comment (dianabarsan) [reply shared-libs/user-management/src/users.js:413]: I'm not sure what the benefit is to create a new array in your suggestion.

File: api/src/controllers/users.js
Comment (dianabarsan) [reply api/src/controllers/users.js:290]: I think this is adds a bit of overhead, because we don't have an authorization library available to us in the users shared library. It's doable, but adds some refactors.

File: api/src/services/authorization.js
Comment (dianabarsan) [reply api/src/services/authorization.js:188]: I think I just copied old code here, because sonar complained.

File: shared-libs/user-management/src/users.js
Comment (dianabarsan) [reply shared-libs/user-management/src/users.js:598]: Sonar always complains.

**Comment by latin-panda** [comment]
Please feel free to dismiss my review to not block this PR 🙂

File: api/src/controllers/users.js
Comment (dianabarsan) [reply api/src/controllers/users.js:290]: Done

File: tests/integration/api/controllers/users.spec.js
Comment (dianabarsan) [reply tests/integration/api/controllers/users.spec.js:1]: I've re-enabled the tests.

**Comment by Benmuiruri** [comment]
Hi @dianabarsan Jenni's [comment](https://github.com/medic/cht-core/pull/9126#issuecomment-2124127168) suggests we should not change the `URL` for [create-user to v3](https://github.com/medic/cht-core/blob/fc98a76f9128b4737a366f210a1330229529474a/admin/src/js/services/create-user.js#L4) and instead use [createMultipleUsers](https://github.com/medic/cht-core/blob/fc98a76f9128b4737a366f210a1330229529474a/admin/src/js/services/create-user.js#L45C13-L45C32) . 

However, that's not correct since the createMultipleUsers is for bulk uploading users... just a heads up on that.

**Comment by dianabarsan** [comment]
Can you send the bulk request to v2 and the single request to v3?

**Review by Benmuiruri — APPROVED** [review]
For the most part all looks good. Just a question, how does the api handle an empty array for the place ? 

There's the create-meta-db  [test](https://github.com/medic/cht-core/blob/master/tests/e2e/default/users/create-meta-db.wdio-spec.js) in my [branch](https://github.com/medic/cht-core/pull/9128) that is returning code 400 with `This user cannot have multiple places` when place is set as an empty `string`

**Comment by dianabarsan** [comment]
Great point @Benmuiruri . I'll handle this case as well.

**Comment by Benmuiruri** [comment]
Hi @dianabarsan I believe this is ready to merge after Jenni and I reviewed it. Please rebase and merge if all is good so that I can rebase and merge Jenni's branches and my branches. Thanks.

**Comment by dianabarsan** [comment]
I think I added one more commit after your last review, that validates that the places list is correct, so we handle:

> There's the create-meta-db [test](https://github.com/medic/cht-core/blob/master/tests/e2e/default/users/create-meta-db.wdio-spec.js) in my https://github.com/medic/cht-core/pull/9128 that is returning code 400 with This user cannot have multiple places when place is set as an empty string

The response is still 400 I believe, but this edge case is handled in code and tested for.

**Review by Benmuiruri — APPROVED** [review]
Thanks for adding the empty facility array  check

**Pull request merged**
